File: CoreTransformer/TransitionPathwayAutoCorrelationMap.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2016 ApexRMS.
'
'************************************************************************************

Imports SyncroSim.Core

Class TransitionPathwayAutoCorrelationMap
    Inherits STSimMapBase3(Of TransitionPathwayAutoCorrelation)

    Public Sub New(
        ByVal scenario As Scenario,
        ByVal collection As TransitionPathwayAutoCorrelationCollection)

        MyBase.New(scenario)

        For Each Item As TransitionPathwayAutoCorrelation In collection
            Me.TryAddItem(Item)
        Next

    End Sub

    Public Function GetCorrelation(
        ByVal transitionGroupId As Integer,
        ByVal stratumId As Integer,
        ByVal secondaryStratumId As Nullable(Of Integer),
        ByVal iteration As Integer,
        ByVal timestep As Integer) As TransitionPathwayAutoCorrelation

        Return Me.GetItem(
            transitionGroupId,
            stratumId,
            secondaryStratumId,
            iteration,
            timestep)

    End Function

    Private Sub TryAddItem(ByVal item As TransitionPathwayAutoCorrelation)

        Try

            Me.AddItem(
                item.TransitionGroupId,
                item.StratumId,
                item.SecondaryStratumId,
                item.Iteration,
                item.Timestep,
                item)

        Catch ex As STSimMapDuplicateItemException

            Dim template As String =
                "A duplicate transition pathway auto-correlation was detected: More information:" & vbCrLf &
                "Transition Group={0}, {1}={2}, {3}={4}, Iteration={5}, Timestep={6}."

            ExceptionUtils.ThrowArgumentException(
                template,
                Me.GetTransitionGroupName(item.TransitionGroupId),
                Me.PrimaryStratumLabel,
                Me.GetStratumName(item.StratumId),
                Me.SecondaryStratumLabel,
                Me.GetSecondaryStratumName(item.SecondaryStratumId),
                STSimMapBase.FormatValue(item.Iteration),
                STSimMapBase.FormatValue(item.Timestep))

        End Try

    End Sub

End Class

File: src/Runtime/TransitionPathwayAutoCorrelationMap.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2021 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using SyncroSim.Core;

namespace SyncroSim.STSim
{
    internal class TransitionPathwayAutoCorrelationMap : STSimMapBase4<TransitionPathwayAutoCorrelation>
    {
        public TransitionPathwayAutoCorrelationMap(Scenario scenario, TransitionPathwayAutoCorrelationCollection collection) : base(scenario)
        {
            foreach (TransitionPathwayAutoCorrelation Item in collection)
            {
                this.TryAddItem(Item);
            }
        }

        public TransitionPathwayAutoCorrelation GetCorrelation(int transitionGroupId, int stratumId, int? secondaryStratumId, int? tertiaryStratumId, int iteration, int timestep)
        {
            return this.GetItem(transitionGroupId, stratumId, secondaryStratumId, tertiaryStratumId, iteration, timestep);
        }

        private void TryAddItem(TransitionPathwayAutoCorrelation item)
        {
            try
            {
                this.AddItem(item.TransitionGroupId, item.StratumId, item.SecondaryStratumId, item.TertiaryStratumId, item.Iteration, item.Timestep, item);
            }
            catch (STSimMapDuplicateItemException)
            {
                string template = "A duplicate transition pathway auto-correlation was detected: More information:" + Environment.NewLine + "Transition Group={0}, {1}={2}, {3}={4}, {5}={6}, Iteration={7}, Timestep={8}.";
                ExceptionUtils.ThrowArgumentException(template, this.GetTransitionGroupName(item.TransitionGroupId), this.PrimaryStratumLabel, this.GetStratumName(item.StratumId), this.SecondaryStratumLabel, this.GetSecondaryStratumName(item.SecondaryStratumId), this.TertiaryStratumLabel, this.GetTertiaryStratumName(item.TertiaryStratumId), STSimMapBase.FormatValue(item.Iteration), STSimMapBase.FormatValue(item.Timestep));
            }
        }
    }
}
